packages/backend/src/core/totalSupply/providers/ArbitrumTotalSupplyProvider.ts

```diff
@@ -0,0 +1,127 @@
+import {
+  assert,
+  Bytes,
+  ChainId,
+  EthereumAddress,
+  UnixTime,
+} from '@l2beat/shared-pure'
+import { BigNumber, utils } from 'ethers'
+
+import { ArbitrumMulticallClient } from '../../../peripherals/arbitrum/multicall/ArbitrumMulticall'
+import { TotalSupplyRecord } from '../../../peripherals/database/TotalSupplyRepository'
+import { EthereumClient } from '../../../peripherals/ethereum/EthereumClient'
+import { MulticallRequest } from '../../../peripherals/ethereum/MulticallClient'
+import { TotalSupplyProvider, TotalSupplyQuery } from '../TotalSupplyProvider'
+
+const erc20Interface = new utils.Interface([
+  'function totalSupply() view returns (uint256)',
+])
+
+export class ArbitrumTotalSupplyProvider implements TotalSupplyProvider {
+  private readonly chainId = ChainId.ARBITRUM
+
+  constructor(
+    private readonly arbitrumClient: EthereumClient,
+    private readonly multicallClient: ArbitrumMulticallClient,
+  ) {
+    assert(
+      this.chainId === this.multicallClient.getChainId(),
+      'chainId mismatch between multicallClient and balance provider',
+    )
+  }
+
+  public getChainId(): ChainId {
+    return this.chainId
+  }
+
+  public async getTotalSupplies(
+    totalSupplyQueries: TotalSupplyQuery[],
+    timestamp: UnixTime,
+    blockNumber: number,
+  ): Promise<TotalSupplyRecord[]> {
+    const canMulticallBeUsed = this.multicallClient.canBeUsed(blockNumber)
+
+    if (canMulticallBeUsed) {
+      return this.fetchUsingMulticall(
+        totalSupplyQueries,
+        timestamp,
+        blockNumber,
+      )
+    }
+
+    return this.fetchUsingPlainCalls(totalSupplyQueries, timestamp, blockNumber)
+  }
+
+  private async fetchUsingMulticall(
+    totalSupplyQueries: TotalSupplyQuery[],
+    timestamp: UnixTime,
+    blockNumber: number,
+  ): Promise<TotalSupplyRecord[]> {
+    const calls = totalSupplyQueries.map(({ tokenAddress }) =>
+      encodeErc20TotalSupplyQuery(tokenAddress),
+    )
+
+    const totalSupplyResponses = await this.multicallClient.multicall(
+      calls,
+      blockNumber,
+    )
+
+    return totalSupplyResponses.map((totalSupplyResponse, i) => {
+      const totalSupply = totalSupplyResponse.success
+        ? decodeErc20TotalSupplyQuery(totalSupplyResponse.data)
+        : 0n
```

Comment:
maybe we should throw an error in such case? wouldn't the 0n be false positive response?

Reply:
https://github.com/l2beat/l2beat/pull/1664/commits/b299256d36c32200625cf8f5f47e1fdd018a4a60

Reply:
Also there is one similar issues regarding balances. 

False-positive fallback for eth balances: https://github.com/l2beat/l2beat/blob/73cc39ab4c64603ed7a9d197e6dbdaabba271575/packages/backend/src/core/balances/providers/EthereumBalanceProvider.ts#L30

False-positive fallback for arb balances:
https://github.com/l2beat/l2beat/blob/73cc39ab4c64603ed7a9d197e6dbdaabba271575/packages/backend/src/core/balances/providers/ArbitrumBalanceProvider.ts#L77

Should we throw instead there too?

Reply:
yes, create a task for that in Linear

Reply:
https://linear.app/l2beat/issue/L2B-1894/determine-what-to-do-when-balance-provider-could-not-obtain-one